### PR TITLE
Common Amlogic GX Project

### PR DESCRIPTION
**Pull request by superna9999**
This PR aims to add support of all Amlogic GX (S905, S905X, S905D, S912) based products under an unbrella project.

This PR adds the:
- Odroid_C2
- Le Potato
- Khadas VIM
- P212 reference design board
 board using Mainline Linux & U-Boot with software video decoding using Kodi GBM/DRM rendering.

### CODE REVIEW

File: packages/linux-drivers/meson_gx_mali_450/package.mk [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+# r7p0
+PKG_VERSION="fb20cef"
+PKG_URL="https://github.com/jakogut/meson_gx_mali_450/archive/$PKG_VERSION.tar.gz"
+PKG_SOURCE_DIR="meson_gx_mali_450-$PKG_VERSION*"
+PKG_DEPENDS_TARGET="toolchain linux"
+PKG_NEED_UNPACK="$LINUX_DEPENDS"
+PKG_SECTION="driver"
+PKG_SHORTDESC="meson_gx_mali_450: Linux drivers for Mali GPUs found in Amlogic Meson SoCs"
+PKG_LONGDESC="meson_gx_mali_450: Linux drivers for Mali GPUs found in Amlogic Meson SoCs"
+
```
Comment (Raybuntu) [inline packages/linux-drivers/meson_gx_mali_450/package.mk]: PKG_IS_KERNEL_PKG="yes" 
That way we can build it with a static GCC linaro toolchain for 64bit Kernel with 32bit userland.

File: packages/graphics/opengl-meson-aarch64/package.mk [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with LibreELEC.  If not, see <http://www.gnu.org/licenses/>.
+################################################################################
+
+PKG_NAME="opengl-meson-aarch64"
+PKG_ARCH="arm aarch64"
```
Comment (Raybuntu) [inline packages/graphics/opengl-meson-aarch64/package.mk]: What ARCH is the mali blob? It's probably aarch64 and not compatible with arm. Do you have a blob fro arm too?

File: projects/Amlogic_GX/options [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        # arm1176jz-s arm1176jzf-s cortex-a8 cortex-a9 cortex-r4
+        # cortex-r4f cortex-m3 cortex-m1 xscale iwmmxt iwmmxt2 ep9312.
+        #
+        TARGET_CPU="cortex-a53"
+        TARGET_CPU_FLAGS="+crc+fp+simd"
+        ;;
+    esac
+
+  # Bootloader to use (syslinux / u-boot / bcm2835-bootloader)
+    if [ "$DEVICE" = "Odroid_C2" ]; then
```
Comment (Raybuntu) [inline projects/Amlogic_GX/options]: Can you put all those variables under one "if" to simplify.

File: projects/Amlogic_GX/options [hunk 4]
```diff
@@ -0,0 +1,146 @@
+################################################################################
+# setup system defaults
+################################################################################
+
+  # The TARGET_CPU variable controls which processor should be targeted for
+  # generated code.
+    case $TARGET_ARCH in
+      aarch64)
```
Comment (Raybuntu) [inline projects/Amlogic_GX/options]: Would be nice to have support for 32bit userspace like C2, WH, WP2 already have in master:
https://github.com/LibreELEC/LibreELEC.tv/commit/de9d1af69f2e25289be3516fa5b35b238e430ea3#diff-ed30c2b16c5c3d9c1884a81bef426347R24

**Comment by CGarces** [comment]
I'm confused with this PR an the usage of `projects/Amlogic_GX/devices/Odroid_C2/`

Odroid C2 already have his own folder at `projects/Odroid_C2`

Will be Odroid_C2 moved from his own project folder to a device of  Amlogic_GX? (and also the Wetek devices)
Or the  `projects/Amlogic_GX/devices/Odroid_C2/` is just a proof of concept to test the Amlogic_GX project and we will continue use  `projects/Odroid_C2`?

**Comment by Raybuntu** [comment]
@CGarces Currently this is going into the amlogic feature branch. Amlogic_GX project is a start to use mainline Kernel, U-Boot for C2, WH, WP2 etc. Ultimately all those seperate Amlogic projects will be superseded by this new project and they will move to be $DEVICES or Amlogic_GX.

File: packages/graphics/opengl-meson-aarch64/package.mk (hunk 2 above)
Comment (superna9999) [reply packages/graphics/opengl-meson-aarch64/package.mk]: Amlogic provides 32bit mali blobs, but I'm unsure they will work using a 64bit kernel

File: projects/Amlogic_GX/options (hunk 4 above)
Comment (superna9999) [reply projects/Amlogic_GX/options]: It should be worth trying

File: packages/tools/u-boot/package.mk [hunk 5]
```diff
@@ -142,6 +142,18 @@ makeinstall_target() {
         cp -PRv $DISTRO_DIR/$DISTRO/splash/boot-logo.bmp.gz $INSTALL/usr/share/bootloader
       fi
       ;;
+  esac
+
+  case $PROJECT in
```
Comment (superna9999) [inline packages/tools/u-boot/package.mk]: Do yo know a simpler way to manage this ?

I thought about a case "$DEVICE$PROJECT" and "Odroid_C2*" would match in either case

File: projects/Amlogic_GX/options (hunk 3 above)
Comment (superna9999) [reply projects/Amlogic_GX/options]: OK

File: packages/tools/u-boot/package.mk (hunk 5 above)
Comment (zalaare) [reply packages/tools/u-boot/package.mk]: Simpler or easier to follow maybe... case in a case?  Really going to be personal opinion.
`case $DEVICE in`
`Odroid_C2) blah blah ;;`
`*) case $PROJECT in`
`Odroid_C2) blah blah ;;`
`esac`
`esac`

File: packages/linux-drivers/meson_gx_mali_450/package.mk [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with LibreELEC.  If not, see <http://www.gnu.org/licenses/>.
+################################################################################
+
+PKG_NAME="meson_gx_mali_450"
+PKG_ARCH="arm aarch64"
+PKG_LICENSE="GPL"
+PKG_SITE="https://github.com/superna9999/meson_gx_mali_450"
```
Comment (CGarces) [inline packages/linux-drivers/meson_gx_mali_450/package.mk]: PKG_SITE=""https://github.com/jakogut/meson_gx_mali_450/"

Comment (superna9999) [reply packages/linux-drivers/meson_gx_mali_450/package.mk]: I don't understand, jakogut's repo is a fork of mine.

Comment (CGarces) [reply packages/linux-drivers/meson_gx_mali_450/package.mk]: Ahh ok, that explain why PKG_SITE<>PKG_URL.

Comment (superna9999) [reply packages/linux-drivers/meson_gx_mali_450/package.mk]: @CGarces oh, you are right I forgot the change the PKG_URL back when I merged back his changes, sorry my bad

**Comment by adamg-xda** [comment]
this looks really promising and would be a great addition given the popularity of amlogic based devices

File: packages/graphics/opengl-meson-aarch64/system.d/unbind-console.service [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+Description=Unbind framebuffer console
+
+ConditionPathExists=/sys/class/vtconsole/vtcon1/bind
+
+[Service]
+Type=oneshot
+ExecStart=/bin/sh -c 'echo 0 > /sys/class/vtconsole/vtcon1/bind'
+
+[Install]
+WantedBy=graphical.target
```
Comment (evelikov) [inline packages/graphics/opengl-meson-aarch64/system.d/unbind-console.service]: FWIW this process of unbinding is completely generic to DRM drivers, even for the nvidia (non-legacy) driver.  
Perhaps it's worth reworking and dropping the duplicated boilerplate at some point.

**Comment by superna9999** [comment]
Hi All, I updated this PR with full mainline:
- U-boot mainline v2018.01
- Linux 4.14
- Mali GBM
- Minor tweaks for DRM/GBM in kodi
- Moved all the mainline specific packages into the Amlogic_GX project

File: projects/Amlogic_GX/options (hunk 4 above)
Comment (superna9999) [reply projects/Amlogic_GX/options]: 32bit is working fine now !

File: packages/devel/libffi/package.mk [hunk 8]
```diff
@@ -30,7 +30,7 @@ PKG_LONGDESC="The libffi library provides a portable, high level programming int
 PKG_TOOLCHAIN="autotools"
 
 PKG_CONFIGURE_OPTS_TARGET="--disable-debug \
-                           --enable-static --disable-shared \
+                           --enable-static --enable-shared \
```
Comment (superna9999) [inline packages/devel/libffi/package.mk]: This is needed until Amlogic delivers GBM only libMali for 32bit... we have the correct binary for 64bit already.

**Pull request merged**